Add Terms Query block

mikachan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Part of https://github.com/WordPress/gutenberg/issues/49094.

This adds a new block called Terms Query, similar to the Query block but for terms rather than posts. It is designed to contain a new Terms Template block, which holds inner blocks with term data for displaying each term.

<!-- In a few words, what is the PR actually doing? -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

To allow for more options and flexibility when displaying terms.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

This block allows for inner blocks and has a wider potential for extensibility compared to the Categories/Terms List block.

I've tried to keep the implementation as simple as possible so we can build up the functionality as we add the other blocks. For the inspector controls, I've tried to stay as close as possible to the options provided by the Categories List block.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Ensure experimental blocks are enabled.
2. Insert the Terms Query block in a template.
3. Ensure the inspector controls render correctly.
4. Ensure the template can be saved successfully with the block.
5. Ensure the expected term data is displayed in the Editor and on the front end.

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

<img width="1508" height="754" alt="image" src="https://github.com/user-attachments/assets/3cccebe3-79e4-4658-845c-513485ac9ba2" />

mikachan committed:
Add Terms Query block

mikachan committed:
Create test fixtures

mikachan committed:
Make block experimental

mikachan committed:
Update docs

mikachan committed:
Update test fixture

mikachan committed:
Fix render_block_core_terms_query

mikachan committed:
Fix render_block_core_terms_query

mikachan committed:
Use termQuery and termQueryId

mikachan committed:
Remove separate showOnlyTopLevel option

mikachan committed:
Merge branch 'trunk' into add/terms-query-block-container

mikachan committed:
Update tests

mikachan committed:
Update parsed fixture

mikachan committed:
Update test fixture

mikachan committed:
Add Core backport PR ref

mikachan committed:
Update readme example

mikachan committed:
Remove term count styling for now

mikachan committed:
Remove editorStyle

scruffian commented on packages/block-library/src/terms-query/terms-query-content.js:
```diff
@@ -0,0 +1,92 @@
+
+	return (
+		<>
+			<InspectorControls>
```
Why not put `InspectorControls` inside the `TermsQueryInspectorControls` component?

mikachan replied:
Good spot, thanks! I've moved both `InspectorControls` into `TermsQueryInspectorControls` in https://github.com/WordPress/gutenberg/pull/70720/commits/96d23958e48b3c3d012349e3310a68ba6d8a11d8.

mikachan committed:
Move all inspector controls into TermsQueryInspectorControls

mikachan committed:
Merge branch 'trunk' into add/terms-query-block-container

mikachan committed:
Add Term Template block (#70747)

* Add term template block

* Add experimental flag

* Fix server sider rendering

* Fix render_block_core_term_template

* Update query to termQuery

* Remove termType

* Fix spacing

* Remove showOnlyTopLevel

* Fix linting

* Create test fixtures

* Remove editorStyle

* Remove clsx

* Add comment for memo

* Add custom loading placeholder

* Add isActiveTerm function

* Handle children filtering on query level

* Disable layout controls

* Fix front end styling

Co-authored-by: mikachan <mikachan@git.wordpress.org>
Co-authored-by: cr0ybot <cr0ybot@git.wordpress.org>
Co-authored-by: fabiankaegy <fabiankaegy@git.wordpress.org>
Co-authored-by: scruffian <scruffian@git.wordpress.org>

mikachan committed:
Add Term Template inner blocks (#71122)

* Add block bindings source

* Use block bindings data as inner blocks

* Rename block_core_term_template_process_block_bindings

* Linting

* Remove term-description support

* Add more term properties

* Apply patch

Co-Authored-By: Bernie Reiter <96308+ockham@users.noreply.github.com>

* Update packages/block-library/src/term-template/index.php

Co-authored-by: Ben Dwyer <ben@scruffian.com>

* Fix class being added twice

* Add default gap to group block

---------

Co-authored-by: Bernie Reiter <96308+ockham@users.noreply.github.com>
Co-authored-by: Ben Dwyer <ben@scruffian.com>
Co-authored-by: mikachan <mikachan@git.wordpress.org>
Co-authored-by: ockham <bernhard-reiter@git.wordpress.org>
Co-authored-by: scruffian <scruffian@git.wordpress.org>

kmanijak commented on packages/block-library/src/term-template/index.php:
```diff
@@ -0,0 +1,218 @@
+/**
+ * Renders the `core/term-template` block on the server.
+ *
+ * @since 6.x.x
```
```suggestion
 * @since 6.9.0
```

And there's multiple other places to address it too.

mikachan replied:
I wasn't sure whether to commit to 6.9.0, but we can always change it if it doesn't make it into the release. I've updated all these references in https://github.com/WordPress/gutenberg/pull/70720/commits/9294f16ee504d6b10a0402b233b4512da4745fbf.

kmanijak commented on packages/editor/src/bindings/term-data.js:
```diff
@@ -0,0 +1,156 @@
+			label: __( 'Count' ),
+			value: termDataValues?.count
+				? `(${ termDataValues.count })`
+				: `(${ termDataValues?.count })`,
```
What's the purpose of this ternany? It seems redundant 🤔

mikachan replied:
Good spot, I think I was doing something different here originally, maybe with the brackets. I've removed this, thanks!

kmanijak commented on packages/block-library/src/term-template/edit.js:
```diff
@@ -0,0 +1,340 @@
+		hide_empty: hideEmpty,
+	};
+
+	const { records: terms, isResolving } = useEntityRecords(
+		'taxonomy',
+		taxonomy,
+		queryArgs
+	);
```
It returns 10 results by default which creates discrepancy between Editor (10 results) and frontend (unlimited results).

1. Have you considered fetching all of the records in editor as well? Or adding a notification that this is just a preview and all of them will be fetched on the frontend (this is not ideal)?
2. There may be a lot of taxonomies sometimes. For practical reasons I'd expect a control to limit the number of results displayed (that would also address the previous point).

mikachan replied:
Ah thanks! There's already a limit of 100 set on the front end, so I've updated this so that the limit is the same in the editor as well.

> There may be a lot of taxonomies sometimes. For practical reasons I'd expect a control to limit the number of results displayed (that would also address the previous point).

Yes I agree, I was thinking of adding this in a follow-up, but maybe this should be included in this initial PR as we already have some query settings on the container block. I'll add this next.

kmanijak replied:
Yeah, I don't mind if it's added in a followup PR 👍 Also, I can implement it if that's helpful!

kmanijak replied:
Covered in https://github.com/WordPress/gutenberg/pull/71358

kmanijak commented on packages/block-library/src/term-template/edit.js:
```diff
@@ -0,0 +1,340 @@
+
+			<div { ...blockProps }>
+				<ul>{ renderTerms() }</ul>
+			</div>
```
Feels like attaching block props to `ul` directly would make layouting easier.

```suggestion
				<ul { ...blockProps }>{ renderTerms() }</ul>
```

And if so, the same change could be applied in the loading skeleton as well.

mikachan replied:
Yeah that looks better, thanks! Done in https://github.com/WordPress/gutenberg/pull/70720/commits/56e244d1e88a486fbfad98298daecd49339bbc2c.

kmanijak reviewed (commented):
Hey @mikachan! What would it take from user to change the layout to let's say a grid? Would it work as in Query Loop that patterns determine the layout and user cannot change it from Inspector Controls?

I tried applying
```
{ layout: { type: 'grid', columnCount: 3 } }
```
to `term-template` but that doesn't seem to do the trick and Terms Query is always a list 🤔 I feel like allowing to more flexible layout options is crucial in this block.

mikachan committed:
Merge branch 'trunk' into add/terms-query-block-container

mikachan committed:
Use 6.9.0

mikachan committed:
Remove ternary

mikachan committed:
Add perPage param in editor

mikachan committed:
Remove wrapper div

mikachan committed:
Add layout options

mikachan committed:
Add columns to grid layout

mikachan committed:
Use ToggleControlGroup for layout selection

mikachan committed:
Tidy up styles

mikachan committed:
Fix linting

mikachan committed:
Fix more spacing

mikachan committed:
Regenerate test fixture

mikachan committed:
Disable grid layout if show hierarchy is enabled

tyxla commented on packages/block-library/src/term-template/edit.js:
```diff
@@ -0,0 +1,512 @@
+	const columnCount = attributes?.columnCount || 3;
+
+	// Switch to list if hierarchical is true and grid is selected.
+	useMemo( () => {
```
Did you mean to use `useEffect()` instead of `useMemo()` here?

mikachan replied:
Gah, yes, `useEffect` would make more sense here for updating attributes. Fixed in https://github.com/WordPress/gutenberg/pull/70720/commits/208ab8c4443a51796d8bab7b22b6b6e8dd09f611. Thanks!

mikachan committed:
Switch useMemo to useEffect

kmanijak reviewed (changes requested):
Thanks for adding List and Grid layouts!  However, I see it's based on Group block and that's pretty limiting to us while I believe we should utilise `supports.layout` that covers it out of the box!

That would require:
- adding `supports.layout: true` to Term Template
- List / Grid inspector controls would manipulate Term Template attributes

```
{ layout: {
  type: 'grid' | 'flow',
  columnCount: N // handled out of the box without a need for custom CSS
  minimumColumnWidth: N // allows for responsiveness
} }
```
- then `packages/block-library/src/term-template/style.scss` would be redundant (at least most of it)

Also, then columns wouldn't be limited to 2-4 columns but we could provide more flexible grid controls even with responsive design. Actually similar to Query Loop:

Auto (responsive) | Manual (fixed columns)
-- | --
<img width="281" height="210" alt="image" src="https://github.com/user-attachments/assets/1079fd09-a5e8-466d-8822-d9266ac3637b" /> | <img width="276" height="212" alt="image" src="https://github.com/user-attachments/assets/ba9c8d7e-c714-44e7-a00b-28eca23c53a0" />

@mikachan, let me know what you think. If you prefer I'm happy to spin up a PR from this branch addressing it but please confirm first so I don't duplicate the effort! 🙌

mikachan committed:
Try using layout support rather than group blocks

mikachan committed:
Simplify layout controls

mikachan commented:
>  However, I see it's based on Group block and that's pretty limiting to us while I believe we should utilise supports.layout that covers it out of the box!

Ah I see, nice! I've tried this in the latest commits, so there are similar grid controls to the Post Template block:

<img width="1509" height="680" alt="image" src="https://github.com/user-attachments/assets/060ddc08-9960-4fe4-a423-a1a4dd8a0875" />

Would this work better for your use cases?

mikachan committed:
Regenerate fixtures

skorasaurus commented:
Hi, 

I haven't kept in touch as much with Gutenberg lately but just wanted to share thanks especially to @mikachan for working on this! 

I've experimented based on a build of https://github.com/WordPress/gutenberg/pull/70720/commits/755e79f5ad45e5a9d7e17c62a429d7ad72c68c06 (regenerating fixtures was the commit name) 

A couple questions or suggestions, [although understanding that the implementation is to be as simple as possible ](https://github.com/WordPress/gutenberg/pull/70720#issue-3229338811)

- [ ] How to hyperlink (or is this done in https://github.com/WordPress/gutenberg/pull/71376) to be able to hyperlink the Term Query? 

eg 
<img width="516" height="349" alt="highlighting a term's text but no ability to hyperlink the text in its block toolbar" src="https://github.com/user-attachments/assets/dde2f3d9-f6bb-4541-b7a7-3fb74d45b82e" />


- [ ] Consider an option to alter the HTML element used for the (or at least bake in future ; a purely code/filter option is acceptable, and I think preferable, since changing an HTML element can create significant accessibility barriers); would this be done in the term name" block; a relatively related issue about this, for post titles is available at https://github.com/WordPress/gutenberg/issues/30549 


Here are some screenshots of a real world example of a potential redesign for an existing project that I manage: I got this together in about 20 minutes and am really pleased by the results ; **I can't remember the last time that I was impressed by the block editor** before now. 

<img width="1227" height="967" alt="image" src="https://github.com/user-attachments/assets/6f692bda-5765-4444-b014-ae21c3f94d64" />

kmanijak commented on packages/block-library/src/term-template/variations.js:
```diff
@@ -0,0 +1,135 @@
+		attributes: { layout: { type: 'grid', columnCount: 3 } },
+		icon: grid,
+		scope: [ 'block', 'inserter' ],
+		innerBlocks: [
```
Minor but `innerBlocks` are the same in both variations so you can consider extracting them to a variable. Obviously, it's fine keeping them as-is so it's more declarative and easier to modify in a future.

mikachan replied:
Nice idea! I've moved the `innerBlocks` to a variable in https://github.com/WordPress/gutenberg/pull/70720/commits/3a0801a8400716511822a0249ba9df8617ce7db1.

kmanijak reviewed (approved):
Thanks a lot for changes around layout! 

> Would this work better for your use cases?

I think so, yes, but TBH my reasoning was to keep it aligned with Query Loop and utilise as much of existing mechanisms as possible rather than implementing custom solution! So appreciate you took care of it.

I think Terms Query in this PR is in a nice state to be merged! And we can iterate further and add more features! Amazing job @mikachan!

mikachan committed:
Add layout controls to block toolbar

mikachan committed:
Add innerblocks variable to variations

mikachan committed:
Add term template class to frontend

mikachan commented:
@skorasaurus Thanks so much for the feedback! I have tried to keep this as simple as reasonably possible to begin with, but we can continue to iterate even once the initial PR is merged, as it's still an experimental block for now.

> How to hyperlink (or is this done in https://github.com/WordPress/gutenberg/pull/71376) to be able to hyperlink the Term Query?

Yes, I think this will be mostly taken care of by https://github.com/WordPress/gutenberg/pull/71376, although as @cr0ybot  mentions, there are some limitations to what can be customised when we're using block bindings for the content. However, this may mean that the way block bindings work will need to be updated to better fit our term data needs, which is also a good thing as this would likely have a wider impact on where block bindings can be used elsewhere.

> Consider an option to alter the HTML element used for the (or at least bake in future ; a purely code/filter option is acceptable, and I think preferable, since changing an HTML element can create significant accessibility barriers); would this be done in the term name" block

I think because block bindings are only available on specific blocks at the moment, we're limited on how the HTML can be customised. However, if this turns out to be something we want, we could always create a new inner block for the term data, rather than using block bindings, if necessary.

> Here are some screenshots of a real world example of a potential redesign for an existing project that I manage: I got this together in about 20 minutes and am really pleased by the results ; I can't remember the last time that I was impressed by the block editor before now.

I love this, it's looking great! So happy to hear that you got it working in such a short amount of time! 🙌 

> regarding the grid layout, why not implement the toggle the same way that the post-template block does, with block controls? I worry about arbitrary differences between these 2 "query" blocks.

@cr0ybot Yes, good point. I've added the grid/list layout controls to the Term Template toolbar in https://github.com/WordPress/gutenberg/pull/70720/commits/2d3b4f26cfc49467be50f02b60ab5ffc82ede32f, so it's more similar to the Query Loop block. I also like having the layout controls in the inspector controls, as they seem easier to find. Maybe we should also update the Query Loop block to match this.

@kmanijak Thanks for doing more testing! I'm just waiting on the e2e tests to be fixed on trunk, and then I think we should merge https://github.com/WordPress/gutenberg/pull/71358 and https://github.com/WordPress/gutenberg/pull/71376, and then we can get this PR merged. Thanks so much for everyone's help with this! ❤️

mikachan committed:
Merge branch 'trunk' into add/terms-query-block-container

kmanijak committed:
Terms Query: add Max Terms control (#71358)

* Change the default perPage to 10

* Add Max Terms control to Terms Query

* Align the default number everywhere

* Fetch the correct number of terms on the frontend

* Limit the number of terms in editor

* Move the max terms setting below hierarchical as it depends on it

* Improve the comment

* Regenerate fixture

* Improve help text

---------

Co-authored-by: Sarah Norris <sarah@sekai.co.uk>
Co-authored-by: kmanijak <karolmanijak@git.wordpress.org>
Co-authored-by: mikachan <mikachan@git.wordpress.org>

cr0ybot committed:
Terms Query: Add link to term-data block bindings (#71376)

* add link to term block bindings

* provide term-data link block binding server-side

* allow core/buttons block in term template

* docs build

---------

Co-authored-by: Sarah Norris <sarah@sekai.co.uk>
Co-authored-by: cr0ybot <cr0ybot@git.wordpress.org>
Co-authored-by: mikachan <mikachan@git.wordpress.org>

scruffian commented:
There are some design discrepancies between the initial state of the group block and this block:
<img width="699" height="361" alt="Screenshot 2025-09-02 at 16 50 46" src="https://github.com/user-attachments/assets/ffd178f0-463a-4ed4-b17c-7be4f35fbe80" />
I think we can deal with this in a followup once we have some design input.

scruffian commented on docs/reference-guides/core-blocks.md:
```diff
@@ -987,6 +987,28 @@ Display the description of categories, tags and custom taxonomies when viewing a
 -	**Supports:** align (full, wide), color (background, link, text), interactivity (clientNavigation), spacing (margin, padding), typography (fontSize, lineHeight), ~~html~~
 -	**Attributes:** textAlign
 
+## Term Template
```
Should we have consistent plurals? Either `Terms Template` or `Term Query`? Actually maybe this way makes sense since the Query is for all terms but the template is for one.

mikachan replied:
> Actually maybe this way makes sense since the Query is for all terms but the template is for one.

This was the main reason I went with "terms" for the container block and "term" for the inner blocks, as the container is for all terms, but the inner blocks are for each individual term. But every time I've named one of the blocks I think about this 😅 I think this is pretty easy to change later, as the block is still experimental.

scruffian commented:
<img width="288" height="563" alt="Screenshot 2025-09-02 at 16 56 41" src="https://github.com/user-attachments/assets/2e6b07df-8353-45ad-8d3e-b40fa8530228" />
It would be great if "Style" was above "Layout" since the layout settings follow from the style ones....

scruffian commented on packages/block-library/src/terms-query/README.md:
````diff
@@ -0,0 +1,83 @@
+<!-- /wp:terms-query -->
+```
+
+## Block Variations
````
Is this meant to be empty?

mikachan replied:
I don't think that should be there at all.. I've removed it, thanks!

scruffian reviewed (approved):
I left a few nits but nothing that should stop this merging! Great work 👏

mikachan committed:
Add missing class from editor

mikachan committed:
Add help text to layout controls

mikachan committed:
Remove Block Variations readme section

mikachan commented:
> Yea, this is probably one reason why the Query block doesn't include the main layout switch in the sidebar. The styles slot for InspectorControls is rendered after other controls, including ones from block supports.

Oh, good point. Maybe we should remove the layout controls from the inspector controls sidebar for this block then, and if we want to explore adding them back, we can do it in a separate PR for both blocks (Query and Terms Query), and include a way to move this option into the Layout section.

kmanijak commented:
> > It would be great if "Style" was above "Layout" since the layout settings follow from the style ones....
> Yes! I don't know how to reorder these inspector controls. I've tried changing the group to styles, which moves it into the Styles tab, but that's the only option I can find for these controls.

@mikachan , @scruffian here's a PR moving the Styles (List | Grid) above the Layout section. I strongly believe we should keep the Style in Inspector Controls for discoverability reasons: https://github.com/WordPress/gutenberg/pull/71479.

t-hamano commented on packages/block-library/src/term-template/editor.scss:
```diff
@@ -0,0 +1,24 @@
+			margin-bottom: 0.25em;
+			background-color: $gray-100;
+			border-radius: 2px;
+			animation: loadingpulse 1.5s ease-in-out infinite;
```
```suggestion
			@media not ( prefers-reduced-motion ) {
				animation: loadingpulse 1.5s ease-in-out infinite;
			}
```

t-hamano commented on packages/block-library/src/term-template/block.json:
```diff
@@ -0,0 +1,69 @@
+			"style": true
+		}
+	},
+	"allowedBlocks": [
+		"core/paragraph",
+		"core/heading",
+		"core/image",
+		"core/buttons",
+		"core/button",
+		"core/group"
+	],
```
Are there any reasons for restricting blocks?

mikachan replied:
These are the blocks that we're currently able to attach block bindngs to, but I don't think this means we need to limit the allowed blocks to only these. Should we allow all blocks?

t-hamano replied:
Just like with the post template block, I think it would be fine to allow all blocks

mikachan replied:
That makes sense, I've removed `allowedBlocks` in https://github.com/WordPress/gutenberg/pull/70720/commits/972d952080ccbda4fe8fb8e70cdf1901b1b18fd7.

t-hamano commented on packages/block-library/src/term-template/block.json:
```diff
@@ -0,0 +1,69 @@
+{
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 3,
+	"__experimental": true,
```
If we decide now to ship these blocks to core, I think this field is unnecessary. If we're just going to experiment on a Gutenberg PR for a while, I think a core backport PR is unnecessary.

mikachan replied:
It would be great if this block could make it into 6.9, but maybe we could keep it as experimental in Gutenberg up until the last GB version that'll be included before the feature freeze. What do you think?

t-hamano replied:
I think either is fine.

I might have been a bit confused, but a PR to core isn't needed in either case, right? If I recall correctly, the new block will be added automatically when the editor package is synced to core.

mikachan replied:
Yeah you're right, this should come across with the package updates. I'll add the `No Core Sync Required` label.

mikachan replied:
Ah no sorry, I think the new block bindings data will need to be backported! But not the block itself.

t-hamano replied:
Ah yes, we need to add new functions, `wp_register_block_bindings_term_data_source()` and `wp_block_bindings_term_data_get_value` function to the core.

t-hamano commented on packages/block-library/src/terms-query/block.json:
```diff
@@ -0,0 +1,51 @@
+	"supports": {
+		"align": [ "wide", "full" ],
+		"html": false,
+		"layout": true,
```
The `layout` field is not necessary.

mikachan replied:
Removed in https://github.com/WordPress/gutenberg/pull/70720/commits/2a6f7ced21221e1021de3f666f79d1d4ce0476c2.

t-hamano commented on packages/block-library/src/terms-query/inspector-controls.js:
```diff
@@ -0,0 +1,231 @@
+					</ToolsPanelItem>
+
+					<ToolsPanelItem
+						hasValue={ () => termQuery.parent === 0 }
```
```suggestion
						hasValue={ () => termQuery.parent !== 0 }
```

t-hamano commented on packages/block-library/src/terms-query/inspector-controls.js:
```diff
@@ -0,0 +1,231 @@
+					<ToolsPanelItem
+						hasValue={ () => termQuery.parent === 0 }
+						label={ __( 'Show only top level terms' ) }
+						onDeselect={ () => setQuery( { parent: undefined } ) }
```
```suggestion
						onDeselect={ () => setQuery( { parent: 0 } ) }
```

t-hamano commented on packages/block-library/src/terms-query/inspector-controls.js:
```diff
@@ -0,0 +1,231 @@
+								order: 'asc',
+								orderBy: 'name',
+								hideEmpty: true,
+								hierarchical: false,
```
```suggestion
								hierarchical: false,
								parent: 0,
								perPage: 10,
```

t-hamano commented on packages/block-library/src/terms-query/block.json:
```diff
@@ -0,0 +1,51 @@
+			"type": "string"
+		}
+	},
+	"usesContext": [ "templateSlug" ],
```
I think `attributes.namespace` and the `templateSlug` context aren't used anywhere.

mikachan replied:
Removed both of these in https://github.com/WordPress/gutenberg/pull/70720/commits/9101b46beabf560a1ced1cf34f6d7c32273e9494.

t-hamano commented on packages/block-library/src/terms-query/index.php:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	$classnames         = 'wp-block-terms-query';
+	$wrapper_attributes = get_block_wrapper_attributes( array( 'class' => $classnames ) );
```
```suggestion
	$wrapper_attributes = get_block_wrapper_attributes();
```

t-hamano commented on packages/block-library/src/terms-query/README.md:
```diff
@@ -0,0 +1,79 @@
+# Terms Query Block
```
The Query Loop block doesn't have such a README, is it necessary for this block?

mikachan replied:
No, it's probably not needed if the Query Loop block doesn't have one. I've removed it in https://github.com/WordPress/gutenberg/pull/70720/commits/d04f9ac95c66bfbf49f88ebd98e2d1b63a43e5d6.

t-hamano commented on packages/block-library/src/term-template/variations.js:
```diff
@@ -0,0 +1,87 @@
+	],
+];
+
+const variations = [
+	{
+		name: 'list',
+		title: __( 'List' ),
+		description: __( 'Display terms in a list layout.' ),
+		attributes: { layout: { type: 'default' } },
+		isDefault: true,
+		icon: list,
+		scope: [ 'block', 'inserter' ],
+		innerBlocks: createInnerBlocks( __( 'Term Name with Count' ) ),
+	},
+	{
+		name: 'grid',
+		title: __( 'Grid' ),
+		description: __( 'Display terms in a grid layout.' ),
+		attributes: { layout: { type: 'grid', columnCount: 3 } },
+		icon: grid,
+		scope: [ 'block', 'inserter' ],
+		innerBlocks: createInnerBlocks( __( 'Term Card' ), {
+			padding: '1rem',
+		} ),
+	},
+];
```
I personally found it confusing that for the Query Loop block we have to choose the "block inside" variation, but for the Term Query block we have to choose the "layout of the block itself." What about skipping the placeholder?

mikachan replied:
I've tried to keep this functionality similar to the variations that are available for the Post Template block inside the Query Loop block.

> What about skipping the placeholder?

Do you mean removing the variation picker that's shown when the block is inserted? I did this to make it similar to the Post Template block, but maybe we could instead default to the List layout, especially as users can easily change the layout now using the inspector controls.

t-hamano replied:
> maybe we could instead default to the List layout, especially as users can easily change the layout now using the inspector controls.

I prefer this approach. I suspect that not many users will want to use terms in a grid layout 😅

kmanijak replied:
> I prefer this approach. I suspect that not many users will want to use terms in a grid layout 😅

The main use case for this block in WooCommerce is with grid layout. But I don't mind removing the variation picker upon inserting as we'll cover this with attributes upfront anyway.

t-hamano commented on packages/block-library/src/term-template/edit.js:
```diff
@@ -0,0 +1,479 @@
+		return (
+			<div { ...blockProps }>
+				<BlockVariationPicker
+					icon="layout"
```
```suggestion
					icon={ layout }
```
Need to pass an icon SVG and not a string, otherwise the dashicon will be displayed.

t-hamano commented on packages/block-library/src/terms-query/index.php:
```diff
@@ -0,0 +1,48 @@
+function render_block_core_terms_query( $attributes, $content, $block ) {
+	if ( ! isset( $attributes['termQuery'] ) || ! isset( $block ) ) {
+		return '';
+	}
```
Why do we need this conditional statement? At least `$block` cannot be empty.

mikachan replied:
I've removed the `$block` check in https://github.com/WordPress/gutenberg/pull/70720/commits/8fd0ca4ab74089185960385bce9eb55ad0fe5a1c.

t-hamano reviewed (commented):
Thanks for the PR. I've only done a smoke test, but I'll comment on a few things that I noticed.

t-hamano commented:
I know there are multiple PRs targeting this branch, but if we decide to ship this block, can we merge this PR first and then iterate afterwards?

Merging a basic prototype first will allow more contributors to test and contribute to the block. However, breaking changes will not be allowed once merged, so that needs to be addressed in this PR.

However, since 21.6 RC has just been released, we should be able to make breaking changes for the next two weeks until 21.7RC is released.

mikachan committed:
Update packages/block-library/src/term-template/editor.scss

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

mikachan committed:
Update packages/block-library/src/term-template/edit.js

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

mikachan committed:
Update packages/block-library/src/terms-query/inspector-controls.js

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

mikachan committed:
Import layout icon

mikachan committed:
Remove layout field from Terms Query

mikachan committed:
Update packages/block-library/src/terms-query/inspector-controls.js

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

mikachan committed:
Update packages/block-library/src/terms-query/inspector-controls.js

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

mikachan committed:
Update packages/block-library/src/terms-query/index.php

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

mikachan committed:
Remove unused context and attribute

mikachan committed:
Remove Terms Query readme

mikachan committed:
Remove $block conditional

# Conflicts:
#	packages/block-library/src/terms-query/index.php

mikachan committed:
Allow all blocks in Term Template

mikachan committed:
Remove variations

mikachan committed:
Revert "Remove variations"

This reverts commit ab582bf89d6a444c482a940633d156bc7517344e.

mikachan committed:
Add namespace attr back

mikachan committed:
Update term template block.json for variation API

mikachan commented:
@cr0ybot Thanks for the extra context about extending the block using variations, especially the requirements listed [here](https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/extending-the-query-loop-block/#extending-the-block-with-variations). It looks like these requirements are:

- A `namespace` attribute needs to be defined
- Some default variations need to exist for the block
- The `variations` property needs to be present in block.json

Based on this, I've updated the PR with the following:

- Added the `namespace` attribute back
- Added the default variations back in
- Also added a reference to `variations` in the block's block.json file

The main difference between these changes and addressing the feedback from @t-hamano's recent review is keeping the `namespace` attribute and keeping the layout variations.

> FYI, the Woo devs are already taking a look at this as a way to provide better ways for users to create customizable layouts for things like product categories (don't ask me for my source ;]). Without the same extensibility as the query block, this block will be useless to them.

Yes! We've been working closely with the Woo folks to ensure this block provides a good level of extensibility, especially for things like product categories.

t-hamano commented:
> Critically, we must preserve the ability for plugins to provide block variations and patterns that can alter the query of the block as [outlined for the core query block here](https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/extending-the-query-loop-block/). That means, for instance, keeping the `namespace` attribute, among other things.

My understanding is that the `namespace` attribute is a special attribute for the placeholder in the Query Loop blocks with "Choose" and "Start Blank" options, and isn't necessary for a simple block variation picker. Adding the `namespace` attribute now in this PR wouldn't make sense unless there are utility functions like [getTransformedBlocksFromPattern](https://github.com/WordPress/gutenberg/blob/c37da743060dd1d84f56a696a92700f82e67bb0f/packages/block-library/src/query/utils.js#L287) and [useScopedBlockVariations](https://github.com/WordPress/gutenberg/blob/c37da743060dd1d84f56a696a92700f82e67bb0f/packages/block-library/src/query/utils.js#L383).

Additionally, the current icon does not match the rest of the variation picker. This icon needs to be improved anyway.

<img width="668" height="679" alt="image" src="https://github.com/user-attachments/assets/15f4cef9-50ee-4392-ad94-52732d56284b" />

One other thing I noticed: why is the variation picker rendered inside the Term Template block and not the Term Query block?

t-hamano commented:
@cr0ybot if you really want the `namespace` attribute. You can simply filter registerBlockType JS filter or the registe_block_type_args PHP filter, no?

t-hamano commented:
@cr0ybot I understand. However, I believe that we need to investigate the following point before shipping this PR. This behavior differs from the Query Loop block.

> One other thing I noticed: why is the variation picker rendered inside the Term Template block and not the Term Query block?

mikachan committed:
Term Template: Use native layout controls (#71488)

* Try layout.allowSwitching

* Fix block.json

* Fix merge conflict

* Fix imports

* Remove bullets from flex layout

Co-authored-by: mikachan <mikachan@git.wordpress.org>
Co-authored-by: kmanijak <karolmanijak@git.wordpress.org>

mikachan commented:
I've just merged https://github.com/WordPress/gutenberg/pull/71488, which was the only outstanding sub-PR for this branch. This simplifies the layout controls by removing the custom controls and enabling `supports.layout.allowSwitching`.

> One other thing I noticed: why is the variation picker rendered inside the Term Template block and not the Term Query block?

This is because the layout variations are applied to the Term Template block, which I was basing on the Post Template block (inside the Query Loop). But I realise this is different to the Query Loop block - looks like the Post Template variations are all applied as variations of the Query Loop instead.

@cr0ybot Will wait for your PR if you already have one to address this. Thank you! 🙌

t-hamano commented:
> However, breaking changes will not be allowed once merged, so that needs to be addressed in this PR.
> 
> @t-hamano, does it also stand if block was merged as experimental?

Yes. As I understand it, ‘__experimental: true` simply indicates that something will not be shipped into core, not that breaking changes are allowed.

The more stuff we include in your first PR, the higher the risk of introducing changes that break backwards compatibility, so I personally prefer to ship something minimal and robust. That's why I proposed removing the namespace attribute and placeholders.

That said, my understanding is that if we really want to ship a truly experimental block, we can include it in the following settings:

<img width="988" height="159" alt="image" src="https://github.com/user-attachments/assets/c1048cf5-0ef7-440d-96b7-ae2818f0ca53" />

t-hamano commented:
This block, like the Query Loop block, is a bit more complex, so I'd like to ping the core team for visibility @WordPress/gutenberg-core. Any feedback is welcome.

In summary, this PR is shipping two blocks: the Term Query and Term Template blocks. The design is similar to the Query Loop block, but a notable feature is that it uses the Block Bindings API to display term data.

mikachan commented:
>  As I understand it, __experimental: true simply indicates that something will not be shipped into core, not that breaking changes are allowed.

@t-hamano I also didn't realise this, I thought that `experimental: true` meant that the block was put behind the experimental blocks flag, and a user would need to enable experimental blocks to use it. What else do we need to do to ensure this block is behind the correct flag (truly experimental)?

> would you be open to merging "Order" and "Orderby" dropdowns?

@kmanijak Yes, this sounds like a good idea. Are there other components that use a similar pattern to this?

Happy to remove ordering by slug too, sounds good!

mikachan committed:
Fix experimental block setting

mikachan commented:
I think I've figured out the experimental setting issue with this block 👀 Looks like the block was being included in the main array for all blocks rather than the experimental list. Should be fixed with https://github.com/WordPress/gutenberg/pull/70720/commits/6e455694dce268ad16cb9057fb737cb5e1434a7e.

t-hamano commented:
> I think I've figured out the experimental setting issue with this block 👀 Looks like the block was being included in the main array for all blocks rather than the experimental list. Should be fixed with https://github.com/WordPress/gutenberg/pull/70720/commits/6e455694dce268ad16cb9057fb737cb5e1434a7e.

That's right. If we want to include the Term Query block in the block experiment setting, we need to add the block inside the `window?.__experimentalEnableBlockExperiments` if statement.

By the way, I think that we also need to move [the Term Template block](https://github.com/WordPress/gutenberg/pull/70720/files#diff-b26c6c8cc1dd9c9775a53b8f9fcc75cddd2a9a4d1a93846330662edce42e6141R240).

mikachan committed:
Move Term Template to experimental array

mikachan commented:
> That's right. If we want to include the Term Query block in the block experiment setting, we need to add the block inside the window?.__experimentalEnableBlockExperiments if statement.

In my testing, the block is now only showing when experimental blocks are enabled. Are you still seeing the block when the experiment is turned off?

> By the way, I think that we also need to move [the Term Template block](https://github.com/WordPress/gutenberg/pull/70720/files#diff-b26c6c8cc1dd9c9775a53b8f9fcc75cddd2a9a4d1a93846330662edce42e6141R240).

I've moved the Term Template block now as well: https://github.com/WordPress/gutenberg/pull/70720/commits/47423d469762c011719f6d681ba551dfea0efa20.

mikachan committed:
Update test fixtures

ntsekouras commented on packages/block-library/src/term-template/block.json:
```diff
@@ -0,0 +1,73 @@
+		}
+	},
+	"style": "wp-block-term-template",
+	"variations": "file:./variations.js"
```
Is this needed for some reason, since we already pass variations in `index.js`?

mikachan replied:
We were exploring what would allow the block to be easily extended, but we probably don't need this in the block.json file. We can remove it.

ntsekouras commented on packages/block-library/src/term-template/variations.js:
```diff
@@ -0,0 +1,87 @@
+		isDefault: true,
+		icon: list,
+		scope: [ 'block', 'inserter' ],
+		innerBlocks: createInnerBlocks( __( 'Term Name with Count' ) ),
```
Why do we need different metadata in these variations? How is `Term Name with Count` different than `Term Card`?

mikachan replied:
`Term Card` is being used in the grid view as a container for the term details, whereas `Term Name with Count` is more simple inline piece of metadata that includes only the term name and the count. The variations are being reworked in https://github.com/WordPress/gutenberg/pull/71596, so this will probably change before being stabalised.

ntsekouras commented on packages/block-library/src/terms-query/edit.js:
```diff
@@ -0,0 +1,10 @@
+ */
+import TermsQueryContent from './terms-query-content';
+
+const TermsQueryEdit = ( props ) => {
```
Nit: this wrapper doesn't seem needed.

ntsekouras commented on packages/block-library/src/term-template/index.js:
```diff
@@ -0,0 +1,26 @@
+
+export const settings = {
+	icon,
+	variations,
```
How come the variations are here and not in the parent block?

ntsekouras replied:
Are they even needed to be in `inserter` scope too? If they really do, we should have different names and icons as it creates this:

<img width="493" height="437" alt="Screenshot 2025-09-21 at 3 26 14 PM" src="https://github.com/user-attachments/assets/b7c7298f-ce55-49a3-be03-c2c7f69df03d" />

ntsekouras commented on packages/editor/src/bindings/term-data.js:
```diff
@@ -0,0 +1,159 @@
+ * @return {Object} The data fields object.
+ */
+function createDataFields( termDataValues, idValue ) {
+	return {
```
Sorry for asking all these post merge (that might have been already discussed), but why use block binding here?

What are the advantages over dedicated blocks like `term name, term description or whatever` and why do we need all these fields in the binding (ex `link, parent` etc..).

Now that we use paragraphs with bindings, these blocks are affected by global styles of paragraphs.

mikachan replied:
> What are the advantages over dedicated blocks like term name, term description or whatever and why do we need all these fields in the binding (ex link, parent etc..).

We were mainly trying to avoid needing many new inner blocks (and it was good to see how far we could get with block bindings), but as @cr0ybot says, we'll likely end up creating a Term Name block to allow for more flexibility.
